### ISSUE TITLE
CasperJS Sync & Async behavious

### ISSUE THREAD

**Issue by robin-obrien** [n1k0/casperjs#1432]
I really appreciate the work invested by the people behind CasperJS and the ideas behind it are really great. However, one thing that is almost impossible to understand is the flow and dependencies between the steps in the CasperJS (For people that are not part of the core development team). With more than 10 years of programming experience and 7 days working with CasperJS, I have not found documentation explaining how to create any sort of dependency between steps and how to achieve having steps synchronously. I invested many hours of trial and error. Regardless the trial and error, and just from the logic I implemented, Eeventually, my code had so many asyc blocks and scopes that it became not determinant without having something to orchestrate it.

How can I create achieve dependency between steps?

**Comment by robin-obrien** [comment]
One last thing, loops, it is so fundamental having the need to achieve loops, what is the official recommendation achieving loops

**Comment by jefleponot** [comment]
The CasperJS force is to transform any asynchronous event or action into sequential steps (synchronous pseudo). Each step is independent. A timer launch a step when last step was finished.

Loops don't exist with CasperJS. If you want to test many time, you have just to run divers tests scripts or loops in a step.

**Comment by robin-obrien** [comment]
I am not sure I understood, How can you synchronize steps and make some depended on the result of a previous step?

**Comment by istr** [comment]
Like @jefleponot already pointed out steps are synchronous by construction. They are queued and executed in order. If you want to retain some state between steps, use any object in the context of the casperjs test(s) to keep it. Of course you might use any loop construct available to automatically queue similar / parametrized steps.

**Comment by robin-obrien** [comment]
@jefleponot, @istr, I didn't really understand which API is the one I can use to make dependencies between steps, so that a certain step occurs only once another step is performed.

**Comment by istr** [comment]
In CasperJS there is no API that supports branching steps. You have to do it on your own along the lines of https://github.com/n1k0/casperjs/blob/master/tests/suites/casper/steps.js (BTW there you see the looping construct you asked for: each, eachThen). If you want a higher level API, you might want to use something like https://github.com/Huddle/PhantomFlow that adds [Decision Trees](https://en.wikipedia.org/wiki/Decision_tree) on top of CasperJS.

**Comment by robin-obrien** [comment]
Thanks, I'll go over those solutions

**Comment by jefleponot** [comment]
Hi @robin-obrien,

Here is a copy paste oh our main loop.
I hope it will help you.
```
/**
 * Run casper navigation if exists
 *
 * @param  Function  fct     OnCompleted function 
 */
ScenarioClass.prototype.run = function runScenario(fct){
  "use strict";
  var scenario = this;

  this.casper.start().then(function(){
        scenario.start();
  });

  // inscription du scenario      
  this.steps.forEach(function(elm,idx){
    scenario.casper.then(function nextStep(){
      scenario.isRunnig = true;
      scenario.currentStep = idx;
      for (var i=0, l=scenario.stepCallBacks.length;i<l;i++) {
        scenario.stepCallBacks[i].call(scenario.casper, elm, idx);
      }
      return true;
    });
  });

  // inscription des exports
  this.casper.run(function(){
    scenario.stop();
    scenario.complete();
  });
};

```
